File: src/tools/hash.ts
import { z } from "zod";
import { createDVQuery, getDVQueryStatus, getDVEvents } from "../client/rest.js";
import { formatTimeAgo, truncatePath } from "../utils.js";

export const hashLookupSchema = z.object({
  hash: z
    .string()
    .describe("SHA1 (40 chars) or SHA256 (64 chars) hash to hunt across the fleet via Deep Visibility"),
});

function summarizeHashEvent(e: Record<string, any>): string {
  const time = e.eventTime ? formatTimeAgo(e.eventTime) : "unknown";
  const type = e.eventType || "Unknown";
  const process = e.processName || "N/A";
  const agent = e.agentName || "Unknown";
  const user = e.processUser || e.user || "";

  let details = "";
  if (e.filePath) details += ` | ${truncatePath(e.filePath, 60)}`;
  if (e.processCommandLine) details += ` | cmd: ${e.processCommandLine.slice(0, 80)}`;
  if (user) details += ` | ${user}`;

  return `• ${agent} | ${type} | ${process} | ${time}${details}`;
}

export async function handleHashLookup(
  params: z.infer<typeof hashLookupSchema>
) {
  try {
    const hashLength = params.hash.length;
    if (hashLength !== 40 && hashLength !== 64) {
      return {
        content: [
          {
            type: "text" as const,
            text: `Invalid hash format. Expected SHA1 (40 chars) or SHA256 (64 chars), got ${hashLength} chars`,
          },
        ],
        isError: true,
      };
    }
    if (!/^[a-fA-F0-9]+$/.test(params.hash)) {
      return {
        content: [
          {
            type: "text" as const,
            text: "Invalid hash format. Hash must be hexadecimal characters only.",
          },
        ],
        isError: true,
      };
    }

    const hashField = hashLength === 64 ? "SHA256" : "SHA1";
    const dvQuery = `${hashField} = "${params.hash}"`;

    // Last 14 days
    const toDate = new Date().toISOString();
    const fromDate = new Date(Date.now() - 14 * 24 * 60 * 60 * 1000).toISOString();

    // Retry query creation on 409 (S1 limits concurrent DV queries per token)
    let result;
    let createAttempts = 0;
    const maxCreateRetries = 6;

    while (createAttempts < maxCreateRetries) {
      try {
        result = await createDVQuery({
          query: dvQuery,
          fromDate,
          toDate,
        });
        break;
      } catch (createError) {
        const msg = createError instanceof Error ? createError.message : String(createError);
        if (msg.includes("409") && createAttempts < maxCreateRetries - 1) {
          createAttempts++;
          await new Promise((resolve) => setTimeout(resolve, 3000));
          continue;
        }
        throw createError;
      }
    }

    if (!result) {
      return {
        content: [
          {
            type: "text" as const,
            text: "DV query slot busy - another query is still processing. Try again shortly.",
          },
        ],
        isError: true,
      };
    }

    // Poll for completion
    let attempts = 0;
    const maxAttempts = 30;
    let status: { status: string; responseError?: string } = { status: "RUNNING" };

    while (attempts < maxAttempts) {
      await new Promise((resolve) => setTimeout(resolve, 1000));
      status = await getDVQueryStatus(result.queryId);
      if (status.status !== "RUNNING") break;
      attempts++;
    }

    if (status.status === "FAILED") {
      return {
        content: [
          {
            type: "text" as const,
            text: `DV hash query failed: ${status.responseError || "Unknown error"}`,
          },
        ],
        isError: true,
      };
    }

    if (status.status === "RUNNING") {
      return {
        content: [
          {
            type: "text" as const,
            text: `Query still running after ${maxAttempts}s. Use s1_dv_get_events with queryId: ${result.queryId}`,
          },
        ],
      };
    }

    // Fetch events with 409 retry (S1 can report FINISHED before events are ready)
    let events;
    let fetchAttempts = 0;
    const maxFetchRetries = 5;

    while (fetchAttempts < maxFetchRetries) {
      try {
        events = await getDVEvents({
          queryId: result.queryId,
          limit: 50,
        });
        break;
      } catch (fetchError) {
        const msg = fetchError instanceof Error ? fetchError.message : String(fetchError);
        if (msg.includes("409") && fetchAttempts < maxFetchRetries - 1) {
          fetchAttempts++;
          await new Promise((resolve) => setTimeout(resolve, 2000));
          continue;
        }
        throw fetchError;
      }
    }

    if (!events) {
      return {
        content: [
          {
            type: "text" as const,
            text: `Query completed but events not available after retries. Use s1_dv_get_events with queryId: ${result.queryId}`,
          },
        ],
      };
    }

    if (!events.data?.length) {
      return {
        content: [
          {
            type: "text" as const,
            text: `No activity found for ${hashField} ${params.hash} in the last 14 days.`,
          },
        ],
      };
    }

    // Deduplicate by agent to show fleet spread
    const agents = new Set(events.data.map((e: Record<string, any>) => e.agentName || "Unknown"));
    const header = `Hash ${hashField} ${params.hash}\nSeen on ${agents.size} endpoint(s) | ${events.data.length} event(s) in last 14 days:\n\n`;
    const summary = events.data.map(summarizeHashEvent).join("\n");
    const footer = events.pagination?.nextCursor
      ? `\n\n[More results - use s1_dv_get_events with queryId: ${result.queryId}, cursor: ${events.pagination.nextCursor}]`
      : "";

    return {
      content: [
        {
          type: "text" as const,
          text: header + summary + footer,
        },
      ],
    };
  } catch (error) {
    return {
      content: [
        {
          type: "text" as const,
          text: `Error hunting hash: ${error instanceof Error ? error.message : String(error)}`,
        },
      ],
      isError: true,
    };
  }
}

File: tools/hash.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/c0tton-fluff/sentinelone-mcp-server/client"
)

var hexPattern = regexp.MustCompile(`^[a-fA-F0-9]+$`)

var hashReputationTool = ToolDef{
	Name:        "s1_hash_reputation",
	Description: "Look up a hash's reputation verdict and hunt it across the fleet via Deep Visibility. Returns the S1 verdict (malicious/suspicious/unknown) plus endpoints, processes, and file paths where the hash was seen in the last 14 days.",
	InputSchema: map[string]any{
		"type":     "object",
		"required": []string{"hash"},
		"properties": map[string]any{
			"hash": map[string]any{
				"type":        "string",
				"description": "SHA1 (40 chars) or SHA256 (64 chars) hash to look up",
			},
		},
	},
}

func summarizeHashEvent(e map[string]any) string {
	timeStr := "unknown"
	if d := getStr(e, "eventTime"); d != "" {
		timeStr = formatTimeAgo(d)
	} else if d := getStr(e, "createdAt"); d != "" {
		timeStr = formatTimeAgo(d)
	}
	eventType := fallback(getStr(e, "eventType"), "Unknown")
	process := fallback(getStr(e, "processName"), "N/A")
	agent := fallback(getStr(e, "agentName"), "Unknown")

	user := getStr(e, "processUserName")
	if user == "" {
		user = getStr(e, "user")
	}

	var details string
	if fp := getStr(e, "fileFullName"); fp != "" {
		details += " | " + truncatePath(fp, 60)
	}
	if cmd := getStr(e, "processCmd"); cmd != "" {
		details += " | Cmd: " + cmd
	}
	if user != "" {
		details += " | " + user
	}

	return fmt.Sprintf("- %s | %s | %s | %s%s", agent, eventType, process, timeStr, details)
}

func handleHashReputation(ctx context.Context, args json.RawMessage) ToolResult {
	var p struct {
		Hash string `json:"hash"`
	}
	if len(args) > 0 {
		if err := json.Unmarshal(args, &p); err != nil {
			return toolError(fmt.Sprintf("invalid arguments: %v", err))
		}
	}
	if p.Hash == "" {
		return toolError("hash is required")
	}

	if len(p.Hash) != 40 && len(p.Hash) != 64 {
		return toolError(
			fmt.Sprintf("Invalid hash format. Expected SHA1 (40 chars) or SHA256 (64 chars), got %d chars", len(p.Hash)),
		)
	}
	if !hexPattern.MatchString(p.Hash) {
		return toolError("Invalid hash format. Hash must be hexadecimal characters only.")
	}

	// Fast verdict lookup via dedicated API.
	verdict, verdictErr := client.GetHashVerdict(ctx, p.Hash)

	hashField := "SHA1"
	if len(p.Hash) == 64 {
		hashField = "SHA256"
	}
	dvQuery := fmt.Sprintf(`%s = "%s"`, hashField, p.Hash)

	now := time.Now().UTC()
	toDate := now.Format(time.RFC3339)
	fromDate := now.Add(-14 * 24 * time.Hour).Format(time.RFC3339)

	// Retry on 409 (S1 limits concurrent DV queries per token)
	var queryID string
	var err error
	for attempt := range 6 {
		queryID, err = client.CreateDVQuery(ctx, dvQuery, fromDate, toDate, nil, nil)
		if err == nil {
			break
		}
		if !strings.Contains(err.Error(), "409") || attempt == 5 {
			return toolError(fmt.Sprintf("Error hunting hash: %v", err))
		}
		select {
		case <-ctx.Done():
			return toolError(fmt.Sprintf("Error hunting hash: %v", ctx.Err()))
		case <-time.After(3 * time.Second):
		}
	}

	if queryID == "" {
		return toolError("DV query slot busy - another query is still processing. Try again shortly.")
	}

	status, err := pollDVQuery(ctx, queryID)
	if err != nil {
		return toolError(fmt.Sprintf("Error hunting hash: %v", err))
	}

	switch status.Status {
	case "FAILED", "FAILED_CLIENT", "ERROR":
		return toolError(fmt.Sprintf("DV hash query failed: %s", fallback(status.ResponseError, "Unknown error")))
	case "QUERY_CANCELLED":
		return toolError("DV hash query was canceled")
	case "TIMED_OUT", "QUERY_EXPIRED":
		return toolError(fmt.Sprintf("DV hash query %s", status.Status))
	case "FINISHED":
		// continue to fetch events below
	default:
		return toolText(fmt.Sprintf("Query still running after 30s. Use s1_dv_get_events with queryId: %s", queryID))
	}

	// Fetch events with 409 retry
	var events *client.PaginatedResponse
	for attempt := range 5 {
		events, err = client.GetDVEvents(ctx, queryID, 50, "")
		if err == nil {
			break
		}
		if !strings.Contains(err.Error(), "409") || attempt == 4 {
			return toolError(fmt.Sprintf("Error hunting hash: %v", err))
		}
		select {
		case <-ctx.Done():
			return toolError(fmt.Sprintf("Error hunting hash: %v", ctx.Err()))
		case <-time.After(2 * time.Second):
		}
	}

	if events == nil {
		return toolText(fmt.Sprintf("Query completed but events not available after retries. Use s1_dv_get_events with queryId: %s", queryID))
	}

	verdictLine := ""
	if verdictErr == nil && verdict != "" {
		verdictLine = fmt.Sprintf("Verdict: %s\n", verdict)
	}

	if len(events.Data) == 0 {
		return toolText(fmt.Sprintf("%sNo activity found for %s %s in the last 14 days.", verdictLine, hashField, p.Hash))
	}

	// Deduplicate by agent to show fleet spread
	agents := make(map[string]struct{})
	for _, e := range events.Data {
		agents[fallback(getStr(e, "agentName"), "Unknown")] = struct{}{}
	}

	lines := make([]string, len(events.Data))
	for i, e := range events.Data {
		lines[i] = summarizeHashEvent(e)
	}

	header := fmt.Sprintf("%sHash %s %s\nSeen on %d endpoint(s) | %d event(s) in last 14 days:\n\n",
		verdictLine, hashField, p.Hash, len(agents), len(events.Data))
	return toolText(header + strings.Join(lines, "\n"))
}
